src/cdk-experimental/table/virtual-rows.ts

```diff
@@ -0,0 +1,292 @@
+
+  /** Emits when the virtual scroll viewport's CSS `translateY()` value changes. */
+  private _renderedContentOffsetChanged = this._viewport.renderedContentOffsetChanged.pipe(
+      map(() => this._viewport.getOffsetToRenderedContentStart()),
+      filter(offset => offset !== null) as OperatorFunction<number|null, number>,
+      distinctUntilChanged());
+
+  /** Emits when the header rows sticky state changes. */
+  private _headerRowStickyUpdates = new Subject<StickyUpdate>();
+
+  /** Emits when the footer rows sticky state changes. */
+  private _footerRowStickyUpdates = new Subject<StickyUpdate>();
+
+  /**
+   * Observable that emits the data source's complete data set. This exists to implement
+   * {@link CdkVirtualScrollRepeater}.
+   */
+  readonly dataStream: Observable<T[] | ReadonlyArray<T>> =
+      this._dataSourceChanges.asObservable()
+      .pipe(
+          switchMap(dataSource => this.getDataStream(dataSource)),
+          shareReplay(1));
+
+  /**
+   * The datasource for the table. The datasource is wrapped in another observable before passing
+   * it to the table component. This getter will return the original datasource, but it will not be
+   * the same instance received by the table.
+   */
+  @Input()
+  get dataSource(): CdkTableDataSourceInput<T> {
+    return this._innerDataSource;
+  }
+  set dataSource(dataSource: CdkTableDataSourceInput<T>) {
+    if (dataSource !== this._innerDataSource) {
+      this._innerDataSource = dataSource;
+      this._dataSourceChanges.next(dataSource);
+      // Wrap the data source in a virtualized list so it only emits a subset of rows as the user
+      // scrolls.
+      this._table.dataSource = this.virtualizeDataSource(dataSource);
+    }
+  }
+
+  /**
+   * The size of the cache used to store unused views. Setting the cache size to `0` will disable
+   * caching.
+   */
+  @Input()
+  get viewCacheSize(): number {
+    return this._viewRepeater.viewCacheSize;
+  }
+  set viewCacheSize(size: number) {
+    this._viewRepeater.viewCacheSize = size;
+  }
+
+  constructor(
+      private readonly _table: CdkTable<T>,
+      private readonly _tableEl: ElementRef<CdkTable<T>>,
+      @Inject(STICKY_POSITIONING_LISTENER) positioningListener: PositioningListenerProxy,
+      @Inject(_VIEW_REPEATER_STRATEGY) protected readonly _viewRepeater: _RecycleViewRepeaterStrategy<T, RenderRow<T>, RowContext<T>>,
+      @Inject(VIRTUAL_SCROLL_STRATEGY) protected readonly _scrollStrategy: VirtualScrollStrategy,
+      @SkipSelf() private readonly _viewport: CdkVirtualScrollViewport) {
+    positioningListener.setListener(this);
+
+    // Force the table to enable `fixedLayout` to prevent column widths from changing as the user
+    // scrolls. This also enables caching in the table's sticky styler which reduces calls to
+    // expensive DOM APIs, such as `getBoundingClientRect()`, and improves overall performance.
+    this._table.fixedLayout = true;
```

Comment:
In theory this could be overwritten later on by the `fixedLayout` input. I don't know how much of an issue that is.

Reply:
I replaced this assignment with a `ngDevMode` error to ensure developers explicitly set `fixedLayout` in their component templates.